### ISSUE TITLE
Question: How did you avoid typescript errors not finding node modules?

### ISSUE THREAD

**Issue by da-mkay** [meltedspark/electron-angular-native#96]
Hi there, nice project! I'm trying to figure out how some things work in your project for better understanding 😄 I hope it's okay to post a question here 👼 .

When importing a node-module in your AppComponent, for example using the following line, it just works:
`import * as fs from 'fs';`

However, if I try to set up an Angular project manually that I can later open in electron, I am not able to compile, because I get the following error:
`error TS2307: Cannot find module 'fs'.`
What I did:
I created a simple new Angular 7 app + added a custom webpack config (listing 'fs' as external as in your webpack config) + updated angular.json to use that config (using @angular-builders/custom-webpack:browser and customWebpackConfig-option, exactly as you did). Then run `ng build`.
NOTE: @types/node and @angular-devkit/build-angular are installed and my editor (VSCode) is not showing any error.

Am I missing something? How do I get the ts-code compiled?

Best regards

**Comment by da-mkay** [comment]
Hi and sorry for the late reply. I did some further investigation, because the settings mentioned in docs (even the performance section) didn't change anything for me. However, I now have full speed over 1-Gbit LAN (Samba, Windows 10 client) 👍 . I will share my new setup here for the case anyone else has the same problems.

My previous setup was the following 🐌  :

- multiple HDDs, each having one partition encrypted with LUKS (options: -c aes-xts-plain64 -s 512 --align-payload 512)
- ext4 on the encrypted volumes (options: -m 0.3 -T huge)
- mergerfs on top of those ext4 filesystems
- shared mergerfs filesystem over smbd
- write files from Windows 10 to samba share

But then I got the mentioned problems:

- heavily varying transfer speed, usually between 20 and 90 MiB/s (average 60 MiB/s), regularly dropping down to 0 MiB/s
- copying of files was often delayed some seconds

So I did dome testing: I tried changing the options of all affected parties, even tried the default options: LUKS, ext4, mergerfs, smbd. However, the problems persisted. Sometimes, a little improvement was noticeable, sometimes not. 
Then I started to remove one part after another from the equation: LUKS, mergerfs, smbd, etc.. 
After removing LUKS I got fullspeed (ext4, mergerfs, smbd). OK, I readded LUKS but removed smbd: I also got fullspeed (copying files directly using dd) etc..
But honestly, neither was an option. But then I remembered, that before using mergerfs, I used mdadm together with XFS, not ext4. So, I switched to XFS and the problems were gone 🎉 .

So my new setup is 🚀 :

- multiple HDDs, each having one partition encrypted with LUKS (options: -c aes-xts-plain64 -s 512)
- XFS on the encrypted volumes (default options)
- mergerfs on top of XFS filesystems (basically default options: defaults,allow_other,direct_io,use_ino,func.getattr=newest,category.create=mfs)
- shared mergerfs filesystem over smbd (no special settings applied, Ubuntu's defaults)

Using this setup I got immediate transfers (no delay) combined with almost constant speed when copying from Windows 10, no drops down to 0 MiB/s.
OK, I didn't get fullspeed at that time, just 70-80 MiB/s.
However, I then replaced the CPU with a slightly faster one and now I get constant 100 MiB/s (Well, HDDs get slower when they are getting full, but that's OK 😋 ).

Now you might say, that the faster CPU alone could have solved the problems. Nope 😃  I restored my previous setup and I were back to heavily varying speed combined with drops and delayed transfers, even with the new CPU installed.

So it looks to me like there are some incompatibilities when using LUKS + ext4 + mergerfs + smbd. Removing one of them often improves or solves the problems, but using all together leads to problems. At least for the two (almost hardware identical) systems that I manage.

**Comment by meltedspark** [comment]
Sounds like _tsconfig.json_ issue. Did you override `typeRoots` by any chance? Maybe Angular build is picking the wrong config?

**Comment by da-mkay** [comment]
Nope, I didn't touch the tsconfig, just using Angular's defaults. I set up a project [here](https://github.com/da-mkay/angular-electron-test/). There are just 2 commits. The first one is the output of `ng new` and the second adds the custom_webpack builder and a custom webpack config declaring `fs` module as external. Running `ng build` shows the error that `fs` is not found. I can fix this by setting `"types": ["node"]` in`tsconfig.app.json`, but that doesn't seem to be right and in your project it's also not set.
